lib/Dialect/Calyx/Transforms/GoInsertion.cpp

```diff
@@ -0,0 +1,86 @@
+//===- GoInsertion.cpp - Go Insertion Pass ----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the Go Insertion pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Calyx/CalyxOps.h"
+#include "circt/Dialect/Calyx/CalyxPasses.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+
+using namespace circt;
+using namespace calyx;
+using namespace mlir;
+
+namespace {
+
+/// Adds the group's "go" signal to the guards of assignments within `group`,
+/// with the exception of the "done" terminator. If the assignment already
+/// has a guard, then the bitwise 'and' is taken of the current guard and the
+/// "go" signal. For example:
+///    ```mlir
+///      %go = calyx.group_go %undef : i1
+///
+///      // Case 1: No Guard
+///      %in = %out : i8
+///      =>
+///      %in = %out, %go ? : i8
+///
+///      // Case 2: Guard
+///      %in = %out, %guard ? : i8
+///      =>
+///      %and = comb.and %guard, %go : i1
+///      %in = %out, %and ? : i8
+///    ```
+static void updateGroupAssignmentGuards(OpBuilder &builder, GroupOp &group,
+                                        GroupGoOp &goOp) {
+  group.walk([&](Operation *op) {
+    if (!isa<AssignOp>(op))
+      return;
+    auto assign = cast<AssignOp>(op);
+    if (assign.guard())
+      // Take the bitwise & of the current guard and the group's go signal.
+      assign->setOperand(
+          2, builder.create<comb::AndOp>(group.getLoc(), assign.guard(), goOp));
+    else
+      assign->insertOperands(2, {goOp});
+  });
+}
+
+struct GoInsertionPass : public GoInsertionBase<GoInsertionPass> {
+  void runOnOperation() override {
+    ComponentOp component = getOperation();
+    OpBuilder builder(component->getRegion(0));
+
+    auto undefinedOp =
+        builder.create<LLVM::UndefOp>(component->getLoc(), builder.getI1Type());
+
+    auto wiresOp = component.getWiresOp();
+    wiresOp.walk([&](Operation *op) {
+      if (!isa<GroupOp>(op))
+        return;
+      auto group = cast<GroupOp>(op);
+      OpBuilder builder(group->getRegion(0));
+      // Since the source of a GroupOp's go signal isn't set until the
+      // the Compile Control pass, use an undefined value.
+      auto goOp = builder.create<GroupGoOp>(group->getLoc(), undefinedOp);
+
+      updateGroupAssignmentGuards(builder, group, goOp);
+    });
+  }
+};
+
+} // namespace
```

Comment:
Style thing, but check out the LLVM guidelines about making anonymous namespaces as small as possible: https://llvm.org/docs/CodingStandards.html#anonymous-namespaces. I don't think the `static` helper or the `runOnFunction` need to be in there.

Reply:
Noted. I've taken the helper out, as well as `runOnFunction`'s definition.